### ISSUE TITLE
DSpark on sm_80 (A800): tool calls malformed ~2/3 of the time — DSML corruption + repetition loops ('0000…') in tool_call arguments; disabling spec recovers 33%→89%

### ISSUE THREAD

**Issue by ZacharyZcR** [allover326/deepseek-v4-cmp170hx#9]
## Summary

With DSpark speculative decoding enabled on sm_80 (A800), **tool calling is unreliable**: roughly 2/3 of `tool_choice=auto` requests come back malformed — corrupted DSML, repetition loops inside `tool_calls` arguments (`0000000…`), or plain-text pseudo tool tags. Plain chat (no `tools`) is unaffected. **Disabling `--speculative-config` raises tool-call success from ~33% to ~89%** on the same instances, same prompts.

Filing here because this repo is the most active sm_80 DSpark lineage we could find; our tree shares it (details below). Happy to run experiments on our hardware if useful.

## Environment

- GPUs: 3 nodes of A800 80GB PCIe (sm_80): 2× nodes with 8 GPUs running 2× TP=4 instances each, 1× node with 4 GPUs (single TP=4 instance) — 5 instances total, all show the issue
- Tree: vLLM `0.23.0` dev, base commit `3843563dd` (lineage matches `haosdent/vllm@dsv4-flash-a100`: contains the dspark speculative config, `v1/worker/gpu/pp_utils.py`, and `v1/worker/gpu/spec_decode/dspark/{speculator,utils}.py`; does **not** contain 0005a's `_prefill_topk_needs_torch_fallback` — predates it)
- Model: `DeepSeek-V4-Flash-0731` (official HF weights, bf16 + block-fp8)
- Launch (per instance):
  ```
  --tensor-parallel-size 4 --enable-expert-parallel
  --tokenizer-mode deepseek_v4 --kv-cache-dtype fp8 --block-size 256
  --max-num-seqs 8 --max-model-len 272000 --max-num-batched-tokens 4096
  --enable-prefix-caching
  --tool-call-parser deepseek_v4 --enable-auto-tool-choice --reasoning-parser deepseek_v4
  --speculative-config '{"method":"dspark","num_speculative_tokens":3,"draft_sample_method":"greedy"}'
  ```
  Env: `VLLM_USE_BREAKABLE_CUDAGRAPH=1`, `VLLM_SM86_NAN_PROBE=0`, `VLLM_SM86_SINK=0`, `VLLM_SM86_FP8_BF16=0`, `PYTHONPATH` pointed at the patched tree, full source build (custom `.so`s present, not `VLLM_USE_PRECOMPILED`).

## Repro

Standard OpenAI request, nothing exotic:

```json
{"model":"DeepSeek-V4-Flash-0731",
 "messages":[{"role":"user","content":"列出当前目录文件用glob工具"}],
 "tools":[{"type":"function","function":{"name":"glob","description":"match files",
   "parameters":{"type":"object","properties":{"pattern":{"type":"string"}},"required":["pattern"]}}}],
 "tool_choice":"auto","max_tokens":100}
```

Run it ~5 times per instance. With DSpark on, per-instance success (structured `tool_calls` returned) was 0/3–2/3 across our 5 instances (~33% aggregate).

## Failure modes observed (with spec on)

1. **Corrupted DSML** (the model's native tool-call format breaks mid-token-sequence):
   ```
   <｜DSML｜invoke="glob">          # missing name=
   <｜DSML｜tool_calls/>            # empty/self-closed
   <｜DSML｜invoke_directory_check> # invented invoke name
   ```
2. **Repetition loops.** In text: `工具必须用调用调用调用用用用调用调用…` (single token repeated to max_tokens). Worse: inside an otherwise-valid structured `tool_calls` **arguments** stream — a 16,890-char `write` arguments had **4,755 chars of literal `0`s (28%), longest run 139 consecutive `0`s**, breaking the JSON. Client hangs receiving a tool call that never becomes valid.
3. **Plain-text pseudo tool tags** (with thinking enabled): assistant content ends with `<bash>ls -la D:/</bash>` instead of any structured call.

## Control experiments

- Same 5 instances, **spec disabled** (only change: drop `--speculative-config`, restart): success 33% → ~89%. A residual ~10% malformed-DSML rate remains even with spec off, so the tail may be model/parser-level, but the dominant failure clearly tracks DSpark.
- Same weights on an **sm_120 build (official vLLM + PR41834, 6× RTX 5090, TP=2/PP=3, no spec)**: 5/5 clean structured tool calls, zero malformed output in extended use.
- Plain chat (no `tools`) is clean in both configurations, long Chinese/code generations included.

## Question

Is this known? The failure smells like the draft/verify path occasionally committing bad greedy-draft runs when the target is emitting DSML structural tokens (and `0`-heavy JSON content) — precisely where a wrong accepted token is most destructive. We didn't find anything in issues/patches here that touches spec-decode acceptance; 0005a and the c3046d1 rebase both look unrelated. Before we start bisecting `spec_decode/dspark/speculator.py` ourselves:

- Have you (or anyone in the 170HX group) run sustained tool-calling / agent traffic with DSpark on and seen this?
- Any reason to expect the c3046d1 rebase to change acceptance behavior?
- `num_speculative_tokens=3` + `draft_sample_method=greedy` — is there a known-safer setting for structured output?

Thanks — the repo docs have been extremely useful for our A800 deployment.

**Comment by ZacharyZcR** [comment]
Follow-up: we read through the spec-decode path in our tree (`3843563dd` lineage) trying to localize this. Sharing what we ruled out and where the remaining suspicion is, in case it helps.

**Ruled out — the verification math itself.** For our config (`draft_sample_method=greedy` → `HAS_DRAFT_LOGITS=False`, stochastic target):

- `_rejection_kernel` takes the Leviathan branch with `draft_log_prob = 0.0` (delta draft, q=1) → accept iff `target_logprob > log(u)`, i.e. accept w.p. p(x). Correct for a deterministic draft.
- `_resample_kernel`'s one-hot residual masks exactly the rejected token (`-inf`) and gumbel-argmaxes the rest; `_insert_resampled_kernel` overwrites the rejection slot and bumps the count. All correct as far as we can tell.

So mathematically every committed token should be target-distributed — the sampler *can't* produce these loops **if its inputs are coherent**. Which points at the data plumbing:

**Remaining suspects — the draft-token trust chain.** `DSparkSpeculator.draft_tokens` is a persistent `[max_num_reqs, n_spec]` buffer indexed by request slot; `combine_sampled_and_draft_tokens` copies it into the flat `input_ids` layout of the *current* step; `RejectionSampler` then reads `input_ids[logits_indices]` back as ground truth of "what was drafted". Any staleness or slot/layout mismatch across engine steps (batch composition changes, chunked prefill interleaving, CUDA-graph replay with persistent buffers — we run the fork's `VLLM_USE_BREAKABLE_CUDAGRAPH=1`) commits a garbage token whenever the target's p(garbage) happens to clear the fixed `u = rand32(seed, pos)` gate. That would be intermittent, concurrency-correlated, and worst exactly in digit runs / DSML structural tokens where target mass is diffuse — matching what we see. The Markov head's anchor read (`prev = input_ids[req_idx * num_query_per_req]`, fixed stride into the persistent query buffer) has the same shape-assumption exposure; a stale `prev` locks the Markov bias onto the previous token, which would explain why loops repeat *the previous character* specifically.

**Two cheap decisive probes we plan to run** (can report back):
1. spec ON + `--enforce-eager`: if corruption disappears → cudagraph/persistent-buffer staleness confirmed.
2. spec ON + `--max-num-seqs 1`: if clean → slot-reuse/concurrency dependence confirmed.

If you have a quicker discriminator (or know the answer already), happy to test on our 5-instance A800 setup.

**Comment by ZacharyZcR** [comment]
Ran the two probes plus one more. Results narrow this down a lot:

| config | tool-call corruption |
|---|---|
| dspark + `greedy` + CUDA graphs (orig) | ~35–65% malformed |
| dspark + `greedy` + `--enforce-eager` | **7/20 malformed — including sequential (single-stream) requests** |
| dspark + `probabilistic` + CUDA graphs | 4/30 malformed (~13%) |
| spec off | ~13% malformed (same base rate) |

Conclusions:

1. **Not CUDA graphs, not concurrency.** Corruption reproduces with `--enforce-eager` on strictly sequential requests. So persistent-buffer staleness / slot reuse (our earlier suspicion) is ruled out.
2. **The excess corruption is specific to `draft_sample_method="greedy"`.** Switching to `probabilistic` drops corruption to the spec-off base rate (~13%, which on this tree appears to be model/parser-level, present without spec too).
3. Things we ruled out by reading the tree (`3843563dd`): the delta-draft acceptance math (`log_q=0`, accept iff `log p > log u`) is correct; `map_draft_to_target` is identity (full-vocab draft, same as the nvidia variant); `tl_rand32` is Triton's Philox `tl.rand`, not a weak hash.

So the bug lives somewhere in what's *different* about the greedy path: `argmax(base_logits + markov_bias)` drafting + one-hot verification, vs gumbel-sampled drafting + full ratio test. For the observed corruption (`<｜DSML｜invoke1">`, `invoke_msg`, `tool_ripples`, long digit runs) to survive verification, `p_target(garbage)` must intermittently be non-tiny at those rows — which smells like the verified logits row occasionally not matching the position being verified, in a way the probabilistic path's stricter `log q` threshold happens to mask. We couldn't pin that without instrumenting a live process (dumping the verified row's top-k at corruption time would settle it).

Practical note for anyone hitting this: `probabilistic` fixes the corruption but measured *slower* than spec-off on our A800s (19.5 vs 23.5 tok/s single-stream — stochastic drafts tank the acceptance rate), so we're running spec-off in production for agent/tool workloads.

**Comment by ZacharyZcR** [comment]
Update: we found vllm-project/vllm@`47a4e410b` (#50183, 2026-08-06) — *"NaN handling in rejection sampler tl.argmax"* — which post-dates this tree's base and describes an OOB-read mechanism that would explain corrupted commits perfectly (all-NaN target logits row → per-block local max all NaN → `tl.argmax` returns an index into the padded region → OOB read in `_compute_global_target_argmax` / `_insert_resampled_kernel` → arbitrary token id committed). Given `VLLM_SM86_NAN_PROBE` exists in this lineage, NaN rows on sm_80 + fp8-kv seemed a very plausible trigger.

We ported the two `NaN→-inf` guards into our tree and re-enabled `greedy` DSpark:

| config | malformed rate |
|---|---|
| greedy, pre-patch | ~35–65% |
| **greedy + #50183 port** | **8/40 = 20%** |
| probabilistic | ~13% |
| spec off (base) | ~13% |

So #50183 is worth having (it fixes a real OOB and measurably helped), but it does **not** fully close the greedy gap on our setup — greedy remains above the base rate, and single-stream throughput with spec was no better than spec-off under today's load, so we've reverted production to spec-off again (keeping the NaN guards in the tree).

Remaining greedy-specific delta after the NaN fix suggests a second mechanism. If you rebase onto a base that includes #50183 (+ #50939 for block verification users), that's one fewer landmine at least.

**Comment by ZacharyZcR** [comment]
Closing the loop with final results: we followed this repo's recipe end-to-end on 2× 8-GPU A800 nodes (TP=4 + EP per instance, not PP) — c3046d1 reconstructed from the tarball (tree SHA verified), patches 0002–0006 applied clean, full source build, plus the two `NaN→-inf` guards from vllm-project/vllm@`47a4e410b` (#50183, which post-dates the c3046d1 base) mounted over `rejection_sampler_utils.py`.

**The tool-call corruption is gone at the source on the new base.** Greedy DSpark `k=5`:

| check | result |
|---|---|
| per-instance tool calls (4 instances × 5) | 20/20 structured `tool_calls`, zero malformed DSML |
| sequential + concurrent battery (10 + 4×5) | 30/30 |
| **streaming** tool calls (the path a proxy can't retry) | 6/6 clean |
| single-stream decode | 23.5 (spec off, old base) → **60.7 tok/s** |

So for anyone else landing here from agent workloads on the old `3843563dd`-era tree: don't patch around it — rebase to c3046d1 per this repo's README and carry #50183.

We'll send a small PR adding the #50183 port as `patches/0007` since the c3046d1 base predates it and it's load-bearing for sm_80 + fp8-kv (NaN rows → `tl.argmax` OOB → arbitrary committed tokens). Thanks again for the meticulous docs — they carried this whole migration.
